#' ---
#' title: "Traces pre-processing and quality control"
#' author: "Isabell Bludau"
#' date: "October 20th, 2018"
#' output:
#'   html_document:
#'     toc: true
#'   html_notebook:
#'     toc: true
#'   pdf_document:
#'     toc: true
#' ---

#' # Overview
#' In this script we ...
#'
#' # Step-by-step workflow
#'
#' ## Load CCprofiler package and set working directory:
library(devtools)
#install_github("CCprofiler/CCprofiler", ref = "DA_module")
#library(CCprofiler)
if (length(grep("nas21.ethz.ch",getwd()))>0) {
  setwd("~/mysonas/CCprofiler")
  load_all()
  setwd("~/mysonas/PRPF8/analysis/output")
  #knitr::opts_knit$set(root.dir = '~/mysonas/PRPF8/analysis/output')
} else {
  setwd("/Volumes/ibludau-1/CCprofiler")
  load_all()
  setwd("/Volumes/ibludau-1/PRPF8/analysis/output")
  knitr::opts_knit$set(root.dir = '/Volumes/ibludau-1/PRPF8/analysis/output')
}

#' ## Load data
traces_list <- readRDS("pepTracesRaw.rda")
design_matrix <- readRDS("design_matrix.rda")
# Subsetting for testing purposes:
#traces_list <- subset(traces_list,
#                 trace_subset_ids = c("P22102","O00468","Q9UBF2",
#                                        "O15067","O75822","Q9Y266",
#                                        "P42167","P61978","O95801",
#                                        "P55060","O14578","O75717","P35221"),
#                 trace_subset_type = "protein_id")

#' ## Reduce traces to a single representative peptide for groups of peptides generated by multiple missed cleavages
#'
#' ##### Integrate traces across conditions to perform the representative peptide selection consistently across conditions
traces_integrated <- integrateTraceIntensities(traces_list,
                                          design_matrix = NULL,
                                          integrate_within = NULL,
                                          aggr_fun = "sum")

#' ##### Annotate the Leading isoform for each peptide
#' Annotate the leading isoform
traces_integrated_iso <- annotateLeadingIsoform(traces_integrated,
                                 isoform_col = "isoform_id",
                                 output_col = "LeadingIsoform")

#' ##### Find the relative position of each peptide in the protein sequence
#' With the leadingIsoform we can use the mapping table that contains the sequences of the
#' fasta file used for the proteomics search to determine the relative position of each peptide
#' within the protein.\
#' Load the mapping table:
map_table <- readRDS("../../data/proteogenomics/HeLa_protgen_combined_1FPKM_header_mapping_v2.rda")
#' Annotate the relative peptide position:
traces_integrated_iso_pos <- annotateRelativePepPos(traces_integrated_iso,
  map_table, multimatch = "first", verbose = F)

#' ##### Select representative peptides
#' Select one peptide as representative of all peptides
#' starting at the same starting position (alternative missed cleavages)
traces_integrated_start <- summarizeAlternativePeptideSequences(
  traces_integrated_iso_pos, topN=1,position="PeptidePositionStart", verbose=F)

#' Select one peptide as representative of all peptides
#' ending at the same end position (alternative missed cleavages)
traces_integrated_start_end <- summarizeAlternativePeptideSequences(
  traces_integrated_start, topN=1,position="PeptidePositionEnd", verbose=F)

#' ##### Subset traces list to selected peptides
traces_list <- subset(traces_list,
  unique(traces_integrated_start_end$traces$id))
common_names <- names(traces_list$minus$trace_annotation)[which(names(traces_list$minus$trace_annotation) %in% names(traces_integrated_start_end$trace_annotation))]
traces_list$minus$trace_annotation <- merge(traces_list$minus$trace_annotation, traces_integrated_start_end$trace_annotation, by=common_names, all.x=T,all.y=F, sort=F)
traces_list$plus$trace_annotation <- merge(traces_list$plus$trace_annotation, traces_integrated_start_end$trace_annotation, by=common_names, sort=F)

#' ## General quality control
#' Align SEC traces across conditions:
alignTraces(traces_list, plot = T, PDF = T)
#' Plot global intensity distributions to estimate relative protein content:
plotGlobalIntensities(traces_list, plot = T, PDF = T)

#' ## Missing value detection and imputation
#' Convert zeros in missing value locations to NA:
pepTracesMV <- findMissingValues(traces_list,
                                 bound_left = 2,
                                 bound_right = 2,
                                 consider_borders = TRUE)

#' Impute NA values by fitting a spline:
pepTracesImp <- imputeMissingVals(pepTracesMV, method = "spline")

#' Plot imputation summary:
plotImputationSummary(pepTracesMV, pepTracesImp, PDF = T, plot_traces = T, max_n_traces = 2)

#' ## Filter by consecutive IDs
pepTracesConsIds <- filterConsecutiveIdStretches(pepTracesImp,
                                             min_stretch_length = 2,
                                             remove_empty = T)
#' Peptide and protein effect of consecutive id filtering:
cat(paste0("Peptides: \n minus: ",length(unique(pepTracesConsIds$minus$trace_annotation$id)),
           " / ", length(unique(pepTracesImp$minus$trace_annotation$id)),
           "\n plus: ",length(unique(pepTracesConsIds$plus$trace_annotation$id)),
           " / ",length(unique(pepTracesImp$plus$trace_annotation$id)),
           "\nProteins: \n minus: ", length(unique(pepTracesConsIds$minus$trace_annotation$protein_id)),
           " / ", length(unique(pepTracesImp$minus$trace_annotation$protein_id)),
           "\n plus: ",length(unique(pepTracesConsIds$plus$trace_annotation$protein_id)),
           " / ", length(unique(pepTracesImp$plus$trace_annotation$protein_id))))

#' ## Calculate Sibling Peptide Correlation
pepTracesImpSPC <- calculateSibPepCorr(pepTracesConsIds,
                                   plot = T, PDF = T)

#' ## Remove proteins with only a single peptide across conditions
minus_multipep <- unique(subset(pepTracesImpSPC$minus$trace_annotation,! is.na(SibPepCorr))$protein_id)
plus_multipep <- unique(subset(pepTracesImpSPC$plus$trace_annotation,! is.na(SibPepCorr))$protein_id)
multipep <- unique(c(minus_multipep,plus_multipep))
pepTracesImpSPCmultipep <- subset(pepTracesImpSPC,trace_subset_ids=multipep,trace_subset_type="protein_id")
#' Peptide and protein effect of single peptide filtering:
cat(paste0("Peptides: \n minus: ",length(unique(pepTracesImpSPCmultipep$minus$trace_annotation$id)),
           " / ", length(unique(pepTracesImpSPC$minus$trace_annotation$id)),
           "\n plus: ",length(unique(pepTracesImpSPCmultipep$plus$trace_annotation$id)),
           " / ",length(unique(pepTracesImpSPC$plus$trace_annotation$id)),
           "\nProteins: \n minus: ", length(unique(pepTracesImpSPCmultipep$minus$trace_annotation$protein_id)),
           " / ", length(unique(pepTracesImpSPC$minus$trace_annotation$protein_id)),
           "\n plus: ",length(unique(pepTracesImpSPCmultipep$plus$trace_annotation$protein_id)),
           " / ", length(unique(pepTracesImpSPC$plus$trace_annotation$protein_id))))

#' ## Update trace annotation
#' Add mock replicate peptide correlation:
# @TODO: Update functions not to depend on this!
pepTracesImpSPCmultipep$minus$trace_annotation$RepPepCorr=1
pepTracesImpSPCmultipep$plus$trace_annotation$RepPepCorr=1
#' Update fraction annotation:
pepTracesImpSPCmultipep <- updateTraces(pepTracesImpSPCmultipep)

#' ## Crate an integrated traces object across all conditions
#' Sum intensities across conditions to enable protein feature finding on consensus traces:
pepTracesImpSPCmultipepSum <- integrateTraceIntensities(pepTracesImpSPCmultipep,
                                          design_matrix = NULL,
                                          integrate_within = NULL,
                                          aggr_fun = "sum")

#' ## Anntate traces with genomic coordinates
#' ##### Combine traces with RNAseq information
#' Format RNAseq mapping table:
map_table$IsoformId <- gsub("\\|.*", "", gsub(">.*?\\|","", map_table$header))
isomap <- unique(map_table[,.(LeadingIsoform=IsoformId, LeadingEnsemblProteinSp=protein)])
isomap <- isomap[!duplicated(LeadingIsoform)] # Take the first protein id if the seq is the same
#' Update trace annotation:
pepTracesImpSPCmultipepSum$trace_annotation <- merge(pepTracesImpSPCmultipepSum$trace_annotation, isomap,by="LeadingIsoform")
pepTracesImpSPCmultipepSum$trace_annotation[, LeadingEnsemblProtein := gsub("_.*", "", LeadingEnsemblProteinSp)]
pepTracesImpSPCmultipepSum$trace_annotation <- pepTracesImpSPCmultipepSum$trace_annotation[order(id)]
#' ##### Load genomic information from the ensembldb package
library(EnsDb.Hsapiens.v86)
ensdb <- EnsDb.Hsapiens.v86

#' ## Annotate the genomic coordinates
#' Annotate summed traces
traces_coordinates <- annotateGenomicCoordinates(pepTracesImpSPCmultipepSum, db=ensdb, proteinIdCol="LeadingEnsemblProtein", verbose=T)
#' Transfer annotation to traces list
traces_coordinates_list <- copy(pepTracesImpSPCmultipep) 
traces_coordinates_list$minus$genomic_coord <- traces_coordinates$genomic_coord[traces_coordinates_list$minus$trace_annotation$id]
traces_coordinates_list$plus$genomic_coord <- traces_coordinates$genomic_coord[traces_coordinates_list$plus$trace_annotation$id]

#' ## Save objects
saveRDS(traces_coordinates_list, "pepTracesImpSPCmultipep.rda")
saveRDS(traces_coordinates,"pepTracesImpSPCmultipepSum.rds")
