#' ## Load data
pepTracesRaw <- readRDS("pepTracesRaw.rda")
design_matrix <- readRDS("design_matrix.rda")
# Subsetting for testing purposes:
#pepTracesRaw <- subset(pepTracesRaw,
#                 trace_subset_ids = c("P22102","O00468","Q9UBF2",
#                                        "O15067","O75822","Q9Y266",
#                                        "P42167","P61978","O95801",
#                                        "P55060","O14578","O75717","P35221"),
#                 trace_subset_type = "protein_id")

#' ## Reduce traces to a single representative peptide for groups of peptides generated by multiple missed cleavages
#'
#' ##### Integrate traces across conditions to perform the representative peptide selection consistently across conditions
traces_integrated <- integrateTraceIntensities(pepTracesRaw,
                                          design_matrix = NULL,
                                          integrate_within = NULL,
                                          aggr_fun = "sum")

#' ##### Annotate the Leading isoform for each peptide
#' Annotate the leading isoform
traces_integrated_iso <- annotateLeadingIsoform(traces_integrated,
                                 isoform_col = "isoform_id",
                                 output_col = "LeadingIsoform")

#' ##### Find the relative position of each peptide in the protein sequence
#' With the leadingIsoform we can use the mapping table that contains the sequences of the
#' fasta file used for the proteomics search to determine the relative position of each peptide
#' within the protein.\
#' Load the mapping table:
map_table <- readRDS("../../PRPF8_protgen_combined_1FPKM_header_mapping_v2.rda")
#' Annotate the relative peptide position:
traces_integrated_iso_pos <- annotateRelativePepPos(traces_integrated_iso,
  map_table, multimatch = "first", verbose = F)

saveRDS(traces_integrated_iso_pos, "traces_integrated_iso_pos.rds")

#' ##### Select representative peptides
#' Select one peptide as representative of all peptides
#' starting at the same starting position (alternative missed cleavages)
traces_integrated_start <- summarizeAlternativePeptideSequences(
  traces_integrated_iso_pos, topN=1,position="PeptidePositionStart", verbose=F)

saveRDS(traces_integrated_start, "traces_integrated_start.rds")

#' Select one peptide as representative of all peptides
#' ending at the same end position (alternative missed cleavages)
traces_integrated_start_end <- summarizeAlternativePeptideSequences(
  traces_integrated_start, topN=1,position="PeptidePositionEnd", verbose=F)

saveRDS(traces_integrated_start_end, "traces_integrated_start_end.rds")

#' ##### Subset traces list to selected peptides
pepTracesRepresentativePeps <- subset(pepTracesRaw,
  unique(traces_integrated_start_end$traces$id))
common_names <- names(pepTracesRepresentativePeps$control$trace_annotation)[
  which(names(pepTracesRepresentativePeps$control$trace_annotation) 
        %in% names(traces_integrated_start_end$trace_annotation))]
pepTracesRepresentativePeps$control$trace_annotation <- 
  merge(pepTracesRepresentativePeps$control$trace_annotation, 
        traces_integrated_start_end$trace_annotation, by=common_names, all.x=T, all.y=F, sort=F)
pepTracesRepresentativePeps$depleted$trace_annotation <- 
  merge(pepTracesRepresentativePeps$depleted$trace_annotation, 
        traces_integrated_start_end$trace_annotation, by=common_names, all.x=T, all.y=F, sort=F)

saveRDS(pepTracesRepresentativePeps, "pepTracesRepresentativePeps.rds")

#' ## Missing value detection and imputation
#' Convert zeros in missing value locations to NA:
pepTracesMV <- findMissingValues(pepTracesRepresentativePeps,
                                 bound_left = 1,
                                 bound_right = 1,
                                 consider_borders = TRUE)

#' Impute NA values by fitting a spline:
pepTracesImp <- imputeMissingVals(pepTracesMV, method = "spline")

#' Plot imputation summary:
plotImputationSummary(pepTracesMV, pepTracesImp, PDF = T, plot_traces = T, max_n_traces = 2)

#' ## Filter by consecutive IDs
pepTracesConsIds <- filterConsecutiveIdStretches(pepTracesImp,
                                             min_stretch_length = 2,
                                             remove_empty = T)

saveRDS(pepTracesConsIds, "pepTracesConsIds.rds")

#' ## Remove single peptide genes
pepTracesConsIds_multiPep <- filterSinglePeptideHits(pepTracesConsIds)

saveRDS(pepTracesConsIds_multiPep, "pepTracesConsIds_multiPep.rds")

#' ## Remove outlier peptides
#' Outlier peptides are defined as peptides that do not have any high correlating
#' sibling peptide. Genes with only a single peptide are automatically removed.
pepTraces_maxCorr <- filterByMaxCorr(pepTracesConsIds_multiPep,
                                  cutoff = 0.8, plot = T, PDF=T, name="maxCorrHist")

#' ## Calculate Sibling Peptide Correlation
pepTraces_maxCorr_SPC <- calculateSibPepCorr(pepTraces_maxCorr,
                                       plot = T, PDF = T,
                                       name = "SibPepCorr_densityplot_afterMaxCorrFilter")

#' Update fraction annotation:
pepTracesList_filtered <- updateTraces(pepTraces_maxCorr_SPC)

#' ## Save objects
saveRDS(pepTracesList_filtered, "pepTracesList_filtered.rda")

#' ## Global stats
traces_names <- c(
  "pepTracesRaw",
  "pepTracesRepresentativePeps",
  "pepTracesConsIds",
  "pepTracesConsIds_multiPep",
  "pepTracesList_filtered"
)

countDT <- data.table(evidence=character(), variable=character(), value=integer(), traces=character())

for (traces in traces_names) {
  #traces <- readRDS(paste0(traces_name,".rda"))
  traces_name <- traces
  traces <- eval(as.name(traces))
  
  getIDs <- getProteogenomicsIds(traces)
  
  genes_uniquecontrol <- length(which(! getIDs$control$genes %in% getIDs$depleted$genes))
  genes_uniquedepleted <- length(which(! getIDs$depleted$genes %in% getIDs$control$genes))
  genes_shared <- length(intersect(getIDs$depleted$genes, getIDs$control$genes))
  
  isoforms_uniquecontrol <- length(which(! getIDs$control$isoforms %in% getIDs$depleted$isoforms))
  isoforms_uniquedepleted <- length(which(! getIDs$depleted$isoforms %in% getIDs$control$isoforms))
  isoforms_shared <- length(intersect(getIDs$depleted$isoforms, getIDs$control$isoforms))
  
  isotypic_uniquecontrol <- length(which(! getIDs$control$isotypic %in% getIDs$depleted$isotypic))
  isotypic_uniquedepleted <- length(which(! getIDs$depleted$isotypic %in% getIDs$control$isotypic))
  isotypic_shared <- length(intersect(getIDs$depleted$isotypic, getIDs$control$isotypic))
  
  peptides_uniquecontrol <- length(which(! getIDs$control$peptides %in% getIDs$depleted$peptides))
  peptides_uniquedepleted <- length(which(! getIDs$depleted$peptides %in% getIDs$control$peptides))
  peptides_shared <- length(intersect(getIDs$depleted$peptides, getIDs$control$peptides))
  
  dt <- data.table(genes=c(genes_shared,genes_uniquecontrol,genes_uniquedepleted),
                   isoforms=c(isoforms_shared,isoforms_uniquecontrol,isoforms_uniquedepleted),
                   isotypic=c(isotypic_shared,isotypic_uniquecontrol,isotypic_uniquedepleted),
                   peptides=c(peptides_shared,peptides_uniquecontrol,peptides_uniquedepleted),
                   evidence = c("shared ","control ","depleted "))
  
  dt.m <- melt(dt,id.vars="evidence")
  dt.m[,traces:=traces_name]
  countDT <- rbind(countDT,dt.m)
}

traces_names_dt <- data.table(
  traces=c("pepTracesRaw",
           "pepTracesRepresentativePeps",
           "pepTracesConsIds",
           "pepTracesConsIds_multiPep",
           "pepTracesList_filtered"), 
  traces_name=c("raw",
                "missed cleavages",
                "consecutive ids",
                "no single peptides",
                "max. correlation")
  )

countDT <- merge(countDT, traces_names_dt, by="traces")

countDT$traces_name <- factor(countDT$traces_name, levels=traces_names_dt$traces_name)
countDT$evidence <- factor(countDT$evidence, levels=c("shared ","control ","depleted "))

pdf("traces_stat_counts.pdf",height=4,width=7)
g <- ggplot(countDT,aes(x=traces_name,y=value,fill=evidence, group=variable)) +
  geom_bar(stat="identity") +
  facet_wrap(~ variable, scales="free",nrow=1) +
  theme_classic() +
  theme(legend.position="bottom") + theme(legend.title = element_blank()) +
  labs(y = "count") +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
print(g)
dev.off()

pdf("traces_stat_counts_noIsoInfo.pdf",height=4,width=7)
countDT_sub <- subset(countDT, variable %in% c("genes","peptides"))
g <- ggplot(countDT_sub,aes(x=traces_name,y=value,fill=evidence, group=variable)) +
  geom_bar(stat="identity") +
  facet_wrap(~ variable, scales="free",nrow=1) +
  theme_classic() +
  theme(legend.position="bottom") + theme(legend.title = element_blank()) +
  labs(y = "count") +
  theme(axis.title.x=element_blank()) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
print(g)
dev.off()

write.table(countDT, "traces_stat_counts.txt", sep="\t",quote = F, col.names = T, row.names = F)

#' ## QC plots
pepTracesList_filtered <- lapply(pepTracesList_filtered, function(p){
  setnames(p$trace_annotation, "Gene_names__(primary_)", "Primary_gene_names")
  p
})
class(pepTracesList_filtered) <- "tracesList"

pdf(paste0("PREPROCESSED_PeptideTraces_","PRPF8",".pdf"), width=4.5, height=4)
test_proteins <- c("Q6P2Q9")
pepTest <- subset(pepTracesList_filtered, trace_subset_ids = test_proteins, trace_subset_type = "protein_id")
plot(pepTest, design_matrix = design_matrix, log = FALSE, legend = T, PDF = F,
     name = paste0("PREPROCESSED_PeptideTraces_","PRPF8"), isoformAnnotation = F, plot = TRUE, highlight = NULL, highlight_col = NULL, colour_by = "Primary_gene_names")
dev.off()

test_proteins <- c("Q15029")
pepTest <- subset(pepTracesList_filtered, trace_subset_ids = test_proteins, trace_subset_type = "protein_id")
plot(pepTest, design_matrix = design_matrix, log = FALSE, legend = T, PDF = TRUE,
     name = paste0("PREPROCESSED_PeptideTraces_","EFTUD2"), isoformAnnotation = F, plot = TRUE, highlight = NULL, highlight_col = NULL, colour_by = "Primary_gene_names")

test_proteins <- c("Q96DI7")
pepTest <- subset(pepTracesList_filtered, trace_subset_ids = test_proteins, trace_subset_type = "protein_id")
plot(pepTest, design_matrix = design_matrix, log = FALSE, legend = T, PDF = TRUE,
     name = paste0("PREPROCESSED_PeptideTraces_","SNRNP40"), isoformAnnotation = F, plot = TRUE, highlight = NULL, highlight_col = NULL, colour_by = "Primary_gene_names")

test_proteins <- c("O75643")
pepTest <- subset(pepTracesList_filtered, trace_subset_ids = test_proteins, trace_subset_type = "protein_id")
plot(pepTest, design_matrix = design_matrix, log = FALSE, legend = T, PDF = TRUE,
     name = paste0("PREPROCESSED_PeptideTraces_","SNRNP200"), isoformAnnotation = F, plot = TRUE, highlight = NULL, highlight_col = NULL, colour_by = "Primary_gene_names")

quant_proteins <- c("Q6P2Q9","Q15029")
#colorMap_v = c("#E69F00", "#56B4E9", "#009E73")
#names(colorMap_v) <- quant_proteins
pepTest <- subset(pepTracesList_filtered, trace_subset_ids = quant_proteins, trace_subset_type = "protein_id")
protTest <- proteinQuantification(pepTest, topN = 1000, keep_less = T)
pdf(paste0("PREPROCESSED_ProteinTraces_","PRPF8interactors",".pdf"), width=4.5, height=4)
plot(protTest, design_matrix = design_matrix, log = FALSE, legend = T, PDF = F,
     name = paste0("PREPROCESSED_ProteinTraces_","PRPF8interactors"), 
     isoformAnnotation = F, plot = TRUE, highlight = NULL, highlight_col = NULL,
     #colorMap = colorMap_v,
     colour_by = "Entry_name")
dev.off()

#' ## General quality control
#' Align SEC traces across conditions:
alignTraces(pepTracesList_filtered, plot = T, PDF = T)
#' Plot global intensity distributions to estimate relative protein content:
plotGlobalIntensities(pepTracesList_filtered, plot = T, PDF = T)

